v3_async_wip/v3_async_wip/iothub_mqtt_client.py

```diff
@@ -0,0 +1,627 @@
+
+        :raises: MQTTError if there is an error sending the MethodResponse
+        :raises: ValueError if the size of the MethodResponse payload is too large
+        """
+        topic = mqtt_topic.get_method_topic_for_publish(
+            method_response.request_id, method_response.status
+        )
+        payload = json.dumps(method_response.payload)
+        logger.debug(
+            "Sending method response to IoTHub... (rid: {})".format(method_response.request_id)
+        )
+        await self._mqtt_client.publish(topic, payload)
+        logger.debug(
+            "Sending method response succeeded (rid: {})".format(method_response.request_id)
+        )
+
+    async def send_twin_patch(self, patch: TwinPatch) -> None:
+        """Send a twin patch to IoTHub
+
+        :param patch: The JSON patch to send
+        :type patch: dict, list, tuple, str, int, float, bool, None
+
+        :raises: MQTTError if there is an error sending the twin patch
+        :raises: ValueError if the size of the the twin patch is too large
+        :raises: CancelledError if enabling twin responses is cancelled by network failure
+        """
+        if not self._twin_responses_enabled:
+            await self._enable_twin_responses()
+
+        request = await self._request_ledger.create_request()
+        topic = mqtt_topic.get_twin_topic_for_publish(
+            method="PATCH", resource_location="/properties/reported", request_id=request.request_id
+        )
+
+        # Send the patch to IoTHub
+        try:
+            logger.debug("Sending twin patch to IoTHub... (rid: {})".format(request.request_id))
+            await self._mqtt_client.publish(topic, json.dumps(patch))
+        except asyncio.CancelledError:
+            logger.warning(
+                "Attempt to send twin patch to IoTHub was cancelled while in flight. It may or may not have been received (rid: {})".format(
+                    request.request_id
+                )
+            )
+            await self._request_ledger.delete_request(request.request_id)
+            raise
+        except Exception:
+            logger.error("Sending twin patch to IoTHub failed (rid: {})".format(request.request_id))
+            await self._request_ledger.delete_request(request.request_id)
+            raise
+
+        # Wait for a response from IoTHub
+        try:
+            logger.debug(
+                "Waiting for response to the twin patch from IoTHub... (rid: {})".format(
+                    request.request_id
+                )
+            )
+            response = await request.get_response()
+        except asyncio.CancelledError:
+            logger.debug(
+                "Attempt to send twin patch to IoTHub was cancelled while waiting for response. If the response arrives, it will be discarded (rid: {})".format(
+                    request.request_id
+                )
+            )
+            await self._request_ledger.delete_request(request.request_id)
```

Comment:
> await self._request_ledger.delete_request(request.request_id) [](http://example.com/codeflow?start=11&length=62)

Same comment as above. You should probably remove the request on all exceptions.